### ISSUE TITLE
`kubectl` should warn user if they attempt to authenticate to the cluster using an expired certificate

### ISSUE THREAD

**Issue by chrissnell** [kubernetes/kubernetes#39767]
**Is this a request for help?** (If yes, you should use our troubleshooting guide and community support channels, see http://kubernetes.io/docs/troubleshooting/.):

**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.):

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):
Feature request

**Kubernetes version** (use `kubectl version`):
v1.5.1

**Environment**:
- **Cloud provider or hardware configuration**: Kubernetes on self-hosted private cloud 
- **OS** (e.g. from /etc/os-release): CoreOS 1235.6.0
- **Kernel** (e.g. `uname -a`):  4.7.3-coreos-r2
- **Install tools**: 
- **Others**:

**What happened**:
When a user with an expired client certificate attempts to use `kubectl` to interact with a cluster, they get a very unhelpful error:
```
%  kubectl get pods
error: You must be logged in to the server (the server has asked for the client to provide credentials)
```
After a lot of digging, the user may discover that their client certificate has expired and needs to be renewed.   

**What you expected to happen**:
I expected to get a more helpful error.   The API service returned a `401 Unauthorized`, yet this is never shown to the user, except when verbosity is bumped up.

I'd like to see two improvements:

* Before attempting to connect to the API service, `kubectl` checks the TLS certificate to see if it the expiration date < current time.  If it has indeed expired, print a helpful message to the user.
* When an API request returns a 401 Unauthorized message, print a more helpful message, perhaps something along the lines of "Server response: client is not authorized to request this resource.  Please verify your authentication credentials and/or certificates for validity". 

**How to reproduce it** (as minimally and precisely as possible):
1. Generate a client certificate that expired before the current system time.
2. Attempt to use it with `kubectl` to interact with the cluster

**Anything else do we need to know**:

**Comment by chrissnell** [comment]
Related but different issue:  #29819

**Comment by alejandroEsc** [comment]
@smarterclayton thockin I have been wondering about this issue that was brought up. I am not sure about asking kubectl to do additional checks on certs, however, the messaging may be a bit better if we include error code and message that are part of the status struct:

```"error: You must be logged in to the server (%d %s: %s)",``` 

What do you think? If so I would like to make the change.

**Comment by chrissnell** [comment]
The certificate expiry check is such a little amount of code and would save users so much headache and time.  Why not let them know?

**Comment by alejandroEsc** [comment]
@chrissnell not debating that fact, if anything i think it would be a nice to have. However, i suspect, and i could be wrong, that this is one case where less information we give people about what is wrong with their auth the more secure we are? I could be wrong and willing to make the change if that is the case.

**Comment by liggitt** [comment]
That's why if we do the expiry check it should be client side

**Comment by chrissnell** [comment]
Agreed @liggitt .  This would be a kubectl feature.  Almost everyone that uses cert-based auth will run into this issue eventually.

**Comment by alejandroEsc** [comment]
@chrissnell @liggitt gotcha, is then an approval to move towards this direction? What additional checks should we propose?

**Comment by chrissnell** [comment]
@alejandroEsc sounds good to me.  I think this could be combined with #29819 to do some basic cert validity checks for the `kubectl` user.  

One design question I have is: should `kubectl` exit with a certain exit code when the cert fails these checks?

**Comment by smarterclayton** [comment]
We do not have custom exit codes today, and if we do add those we would do
it holistically.

**Comment by wangxinnong** [comment]
I don't think we should let kubectl validate the certificate.
It should be the responsibility of api-server to validate the certificate, not the kubectl.

## Cause
The cause of this issue is that when authentication fails, api-server only sets a 401 status and a response body of "Unauthorized" and never returns the authentication error to the caller.

kubectl never uses this body. Instead it generates a fixed error message of "the server has asked for the client to provide credentials".

## Proposal
### api-server
Returns the authentication error in the response body when fail to authenticate to let the caller know what's the cause of authentication failure.

### kubectl
If authentication error occurs, print out the error message returns from api-server's response body.
For example:
error: You must be logged in to the server (Unauthorized: x509: certificate has expired or is not yet valid)

## Issue
Because the authentication is done by a chain of authentication request, it is possible multiply errors will occur, so you will see the following message if a invalid token is used for authentication:
`error: You must be logged in to the server (Unauthorized:  [invalid bearer token, [invalid bearer token, [invalid bearer token, invalid bearer token, malformed JWS, only 1 segments]]])`

## Question
Is there any security issue if the failure cause is returned to the caller?

**Comment by chrissnell** [comment]
It feels like overkill to handle this on the server-side but if that's where everyone wants it, it doesn't matter to me.  

I do urge the team to reconsider the wording of the "You must be logged into the server" message.  This message is totally unhelpful to someone dealing with an auth issue.

**Comment by chrissnell** [comment]
/lifecycle frozen

Still hoping to fix this at some point.

**Comment by enj** [comment]
This needs consensus on where the code should live and exactly what it should do (IIRC that is what kept #87796 from progressing).

**Comment by lmktfy** [comment]
I recommend putting the code in kubectl itself. The warning can be printed regardless of authnz outcome; an expired certificate is always worth warning about.

Same if the certificate is not yet valid. 

So long as kubectl knows the not-before and not-after timestamps, and has access to the current date, it is able to use the existing warning code to inform a user about this.
